state/protocol/badger/mutator.go

```diff
@@ -981,6 +996,73 @@ func (m *FollowerState) epochStatus(block *flow.Header, epochFallbackTriggered b
 
 }
 
+// versionBeaconOnBlockFinalized extracts the VersionBeacons from the parent block's
+// Seals and returns it.
+// This could return multiple VersionBeacons if the parent block contains multiple Seals.
+// The version beacons will be returned in the ascending height order of the seals.
+// Technically only the last seal is relevant.
+func (m *FollowerState) versionBeaconOnBlockFinalized(
+	header *flow.Header,
+) ([]*flow.SealedVersionBeacon, error) {
+	var versionBeacons []*flow.SealedVersionBeacon
+
+	parent, err := m.blocks.ByID(header.ParentID)
+	if err != nil {
+		return nil, fmt.Errorf(
+			"could not get parent (id=%x): %w",
+			header.ParentID,
+			err)
+	}
+
+	seals, err := protocol.OrderedSeals(parent.Payload, m.headers)
+	if err != nil {
+		if errors.Is(err, storage.ErrNotFound) {
+			return nil, fmt.Errorf(
+				"ordering seals: parent payload contains"+
+					" seals for unknown block: %w", err)
+		}
+		return nil, fmt.Errorf("unexpected error ordering seals: %w", err)
+	}
+
+	for _, seal := range seals {
+		result, err := m.results.ByID(seal.ResultID)
+		if err != nil {
+			return nil, fmt.Errorf(
+				"could not retrieve result (id=%x) for seal (id=%x): %w",
+				seal.ResultID,
+				seal.ID(),
+				err)
+		}
+		for _, event := range result.ServiceEvents {
+
+			ev, ok := event.Event.(*flow.VersionBeacon)
+
+			if !ok {
+				// skip other service event types.
+				// validation if this is a known service event type is done elsewhere.
+				continue
+			}
+
+			err := ev.Validate()
+			if err != nil {
+				m.logger.Warn().
+					Err(err).
+					Str("block_id", parent.ID().String()).
+					Interface("event", ev).
+					Msg("invalid VersionBeacon service event")
+				continue
+			}
+
+			versionBeacons = append(versionBeacons, &flow.SealedVersionBeacon{
+				VersionBeacon: ev,
+				SealHeight:    header.Height,
```

Comment:
why it's `header.Height`, not `parent.Height`?

Reply:
Took another look at this. This was a mistake. The problem was that I was referring to the code in PR https://github.com/onflow/flow-go/pull/3736 which was written before https://github.com/onflow/flow-go/pull/4149 which changed how this worked.